### ISSUE TITLE
Creating same size windows on Mac / Windows

### ISSUE THREAD

**Issue by markusmoenig** [rust-windowing/winit#2516]
I have a problem creating exactly the same sized windows for Windows and Mac.

On the Mac I use the LogicalSize for the inner window size and it's perfect. When I use that on Windows 11 the Window is too big (like 1.5 times the size).

When I use PhysicalSize() on the Mac the Window is too small (because of Retina Scale Factor) and when I use it on Windows 11 it is perfect.

How to be able to open the window in the same size on both operating systems ?

Note I am on v0.26 as this is the last version pixels works with.

Thanks

**Comment by madsmtm** [comment]
Hey @markusmoenig, are you sure there's not something else going on in the OS, e.g. some kind of scale factor that is set up somewhere?

Also, could I get you to give me the output of `Window::scale_factor` on both systems?

**Comment by markusmoenig** [comment]
Hi, thanks for the answer.

No, the setup is a very basic winit / pixels window creation, not much going on there. the scale factor is 1 for Windows and 2 for Mac. On Windows it behaves like 1.5.

For now I use PhysicalSize on Window and LogicalSize on macOS, that works but is a bit of a hassle. 

But again I use an older winit version as pixels does not work with the latest ones for some reason.

Thanks,

**Comment by dhardy** [comment]
Is this still the case? Is it just a bug that logical sizes are not scaled correctly on Windows?

Testing shows that X11 (which does not have proper scaling support) still allows (approximately) correct window sizing when specified in logical pixels (scale factor 1.5).

---

> When I use PhysicalSize() on the Mac the Window is too small (because of Retina Scale Factor)

It should be obvious that if you use `PhysicalSize`, you need to handle the scale factor yourself. However, it may not be possible to get the correct scale factor before constructing the window; it is not on Wayland: #3183. As a result, it's impossible to correctly size a *new* window specified via `PhysicalSize` on Wayland: #2275.

**Comment by kchibisov** [comment]
Yes, the `LogicalSize` is the right tool to request a window of the same size regardless of the operating system. So if you want a window of `800x600` points, you use logical, otherwise physical. The only issue you can't do is to constraint the window size you use to create.

**Comment by Ashk3000** [comment]
Hi, I am having an issue where PhysicalSize is being affected by my resolution settings (which I believe is the Mac version of scale). I've included a picture of the menu and screenshots of an app that uses PhysicalSize in fullscreen at different scales. I think apple renders stuff at a higher resolution than your screen. I'm on a 2020 M1 Macbook fyi
<img width="714" alt="Screenshot 2024-12-14 at 8 31 43 PM" src="https://github.com/user-attachments/assets/6f17ffc3-b3b6-4b74-a32c-4757de68364f" />
<img width="1440" alt="Screenshot 2024-12-14 at 8 32 42 PM" src="https://github.com/user-attachments/assets/21ea3340-21ff-40a4-b8cd-772e5282546a" />
<img width="1024" alt="Screenshot 2024-12-14 at 8 34 16 PM" src="https://github.com/user-attachments/assets/1a76fbdd-3e4e-4561-9f0b-ddb278878945" />

![Screenshot 2024-12-14 at 8 33 34 PM](https://github.com/user-attachments/assets/cb2ebdc2-9b39-40bc-858b-145949ab8817)

**Comment by markusmoenig** [comment]
@Ashk3000 I ended up doing something like:

```
        let window = {
            if cfg!(target_os = "macos") {
                let size = LogicalSize::new(width as f64, height as f64);
                WindowBuilder::new()
                    .with_title(window_title)
                    .with_inner_size(size)
                    .with_min_inner_size(size)
                    .with_window_icon(icon)
                    .build(&event_loop)
                    .unwrap()
            } else {
                let size = PhysicalSize::new(width as f64, height as f64);
                WindowBuilder::new()
                    .with_title(window_title)
                    .with_inner_size(size)
                    .with_min_inner_size(size)
                    .with_window_icon(icon) //TODO on Windows
                    .build(&event_loop)
                    .unwrap()
            }
        };
```

I did not see any other way, as both solutions (Logical and Physical) behave differently depending on the platform they are running on, PhysicalSize makes it too small on the Mac, LogicalSize makes it too large on Windows.
